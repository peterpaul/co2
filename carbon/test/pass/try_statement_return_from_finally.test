/*
 * carbon Compiler and programming language for libco2
 *
 * Copyright (C) 2011 Peterpaul Taekele Klein Haneveld
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


["stdio.h"] int printf(char * format, ...);

class Exception
{
	char * message;
}

class AnException : Exception
{
}

int fun (Exception ex)
{
	if (ex != null)
	{
		throw ex;
	}
}

int loop (Exception ex)
{
	printf ("--------------------------------------------------------------------------------\n");
	try {
		printf ("try\n");
		fun (ex);
		return 0;
	} catch (AnException ex) {
		printf ("catch AnException\n");
		delete ex;
		return 1;
	} catch (Exception ex) {
		printf ("catch Exception\n");
		delete ex;
		return 2;
	} finally {
		printf ("finally\n");
		return 3;
	}
	return 4;
}

int main(int argc, char ** argv)
{
	printf("%d\n", loop (null));
	printf("%d\n", loop (new Exception()));
	printf("%d\n", loop (new AnException()));
	return 0;
}