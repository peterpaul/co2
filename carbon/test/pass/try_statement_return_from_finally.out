--------------------------------------------------------------------------------
try
finally
3
--------------------------------------------------------------------------------
try
catch Exception
finally
3
--------------------------------------------------------------------------------
try
catch AnException
finally
3
